waiter/src/waiter/service_description.clj

```diff
@@ -490,6 +490,10 @@
           ; else rely on presence of x-waiter headers to set the run-as-user
           contains-waiter-header? (headers/contains-waiter-header waiter-headers on-the-fly-service-description-keys)
           user-service-description (cond-> sanitized-metadata-description
+                                           (and (contains? (:headers sources) "run-as-user")
```

Comment:
The issue is that they're doing different things, the last case is setting the permitted-user to the current user when there is no permitted-user while this one is setting the permitted-user to the run-as-user (which may be different) unless it was specifically set in another header.

Did the destructuring.